inference/marian-fork/src/common/io_item.h

```diff
@@ -9,19 +9,72 @@ namespace marian {
 namespace io {
 
 struct Item {
-  std::vector<char> bytes;
+  std::shared_ptr<std::vector<char>> bytes;
   const char* ptr{0};
   bool mapped{false};
 
   std::string name;
   Shape shape;
   Type type{Type::float32};
 
+  // Default constructor
+  Item() : bytes(std::make_shared<std::vector<char>>()) {}
+
+  // Copy constructor
+  Item(const Item& other)
+      : bytes(other.bytes),
+        ptr(other.ptr),
+        mapped(other.mapped),
+        name(other.name),
+        shape(other.shape),
+        type(other.type) {}
+
+  // Copy assignment operator
+  Item& operator=(const Item& other) {
+    if(this != &other) {
+      bytes = other.bytes;
+      ptr = other.ptr;
+      mapped = other.mapped;
+      name = other.name;
+      shape = other.shape;
+      type = other.type;
+    }
+    return *this;
+  }
+
+  // Move constructor
+  Item(Item&& other) noexcept
+      : bytes(std::move(other.bytes)),
+        ptr(other.ptr),
+        mapped(other.mapped),
+        name(std::move(other.name)),
+        shape(std::move(other.shape)),
+        type(other.type) {
+    other.ptr = nullptr;
+    other.mapped = false;
+  }
+
+  // Move assignment operator
+  Item& operator=(Item&& other) noexcept {
+    if(this != &other) {
+      bytes = std::move(other.bytes);
+      ptr = other.ptr;
+      mapped = other.mapped;
+      name = std::move(other.name);
+      shape = std::move(other.shape);
+      type = other.type;
+
+      other.ptr = nullptr;
+      other.mapped = false;
+    }
+    return *this;
+  }
```

Comment:
> praise

Thank you for checking for self-assignment, and nulling out the old owner's members.